### DISCUSSION TITLE
Feature proposal: i3 like option to display inactive workspaces that have windows in them

### DISCUSSION THREAD

**Discussion by mobile-ar** [nikitabobko/AeroSpace#844] [feature-ideas] [answered]
I added some functionality I was missing from i3, the ability to see workspaces that are not active, but have at least an open window in them.
This will look like this for someone with 2 monitors:
<img width="183" alt="image" src="https://github.com/user-attachments/assets/d043172a-d897-4931-b328-0cd4e83f391a" />
Here for example I have workspace `G` opened in my main monitor, workspace `S` in the external monitor, and workspaces `M`,`N`, `T`, `X` and `Z` that are grayed out are the ones that have windows in them, but they are not currently active in any monitor.

I implemented this by adding a configuration like this:
``` toml
# Possible values: text|squares|i3 (available from macOS 14 or later)
# - text:        displays menu bar as monospaced text (only active workspaces for each monitor).
# - squares: displays menu bar as square with the workspace name inside. Filled square is the
#                   active monitor, bordered square for other monitors (displays only active workspaces
#                   for each monitor).
# - i3:           same as squares + shows all other workspaces that contains windows in it, like i3.
menu-bar-style = 'text'
```

In this way the user can choose how they want to display this, if they care about the functionality or even if they have the space in the menu bar to display this.

<img width="391" alt="image" src="https://github.com/user-attachments/assets/d776c71f-96ba-41a7-91b2-e514051f60a9" />

Adding here the menu window to better understand from where that is coming.

For consistency in the style, I added the modes to be represented as a circle with the corresponding character of the name
<img width="212" alt="image" src="https://github.com/user-attachments/assets/159c4b84-fc2f-4d1d-9715-ef82000709b3" />

Other modes are supported on the same way, the letter being assigned to the mode will be displayed inside a circle, to differentiate a monitor from a mode.

For example, here I created a mode test 
``` toml
[mode.test.binding]
esc = ['reload-config', 'mode main']
```
that will display like this
<img width="215" alt="image" src="https://github.com/user-attachments/assets/c60e3b5a-560b-43f2-9849-ab5ca97b91fc" />

For the menu icons I'm using native `Image(systemName:)` and icons are from SF Symbols (natively supported).

After further reading of closed/rejected PRs I found this might help solve the issue #56 

This is how it will look in the 3 different configurations in service mode (remember that in normal mode there is no indication of the mode)

Text option for monospaced text as default (parameter value `text` with not monospaced when macOS don't support it): 
<img width="103" alt="image" src="https://github.com/user-attachments/assets/08ae80fd-1fbe-4ab3-9b6a-6245a8ac14dc" />
<img width="87" alt="image" src="https://github.com/user-attachments/assets/3989770f-2ec9-46c2-9d49-d01eecd76ec9" />

Squares uses the SF symbols images to represent each workspace (parameter value `squares`):
<img width="107" alt="image" src="https://github.com/user-attachments/assets/79db1ed4-b21b-4743-841e-2781449909a4" />

i3 style same as squares + display workspaces that have windows (parameter value `i3`):
<img width="212" alt="image" src="https://github.com/user-attachments/assets/159c4b84-fc2f-4d1d-9715-ef82000709b3" />

All this will keep support for emojis (in square and i3 style emojis are rendered without any type of square for legibility)
<img width="199" alt="image" src="https://github.com/user-attachments/assets/14179969-0c20-4a33-b43f-aa706abe555c" />
<img width="204" alt="image" src="https://github.com/user-attachments/assets/f5eecb1d-ff2d-4208-9ab2-32cf580920e1" />

Also as you can see I added the colon `:` to separate visually the mode from the monitors and the pipe `|` to separate the active workspaces/monitors from the background workspaces. This could be removed if it feels too much.

This whole functionality implementation could be fund already developed and working on my fork under https://github.com/mobile-ar/AeroSpace/tree/feature/menu-bar-style-i3-like , it's how I'm currently running Aerospace.

**Comment by nikitabobko** [comment]
Also see: https://github.com/nikitabobko/AeroSpace/issues/711#issuecomment-2541128213

**Reply by mobile-ar** [reply]
I understand now your feeling towards UI, still as humans we all have our own preferences and taste, I think over customization is too much, some people love themes and deep level customization, but I think that this change can improve user interface (UI) and user experience (UX) with an small amount of change and just an extra option of functionality not much more customization that that.

What I suggest here is just a simple way to have the super simple text version currently available, plus adding 1 option for using images in a really simple way* (I will explain below) and 1 option for displaying extra populated but not active workspaces.

*the simple way of displaying images is using the same character assigned to a workspace + the system image you want to display.
For example: for workspace `1` you can use `1.square.fill` when active and `1.square` when inactive. This is using `systemImage` that it’s default system images and this is support from 1 to 0 and from A to Z.
So to achieve this with a simple string interpolation like `“\(workspace.name).square.fill”` and passing that string to `Image(systemName: ...`

1. You have to describe a clear, organized and strict mental model of all possible menubar styles
All possible menu bar styles are 3:
- text: is the actual menu style (default) like ‘[s] *1 | 2
- image: will display the same basic stuff but using system images: a circle with the letter for the modes (like the square brackets in text mode), and square filled for the active workspace (this will be the equivalent of the `*` in text mode), and and square with border for the other monitors currently in display but not active
- image-with-background-workspaces: this will be the same as images (b.) + adding a `|` (pipe) + the list of other workspaces that have windows in them, like ‘| A X Z’ (for example if you have Alacritty in A, Xcode in X and Zed in Z for those workspaces) keeping in mind that the space

2. You have to describe pros and cons of each menubar style:
- text: pro - simple. con: not enough visual help at simple glance
- image: pro - simple, more difference at a glance of the active state of the window, more similar at all other status/menu bar items (usually icon based)
- image-with-background-workspaces: same pros as image. Con: if a user has a window open in each workspace, it could get quite long in the menu (but that might not be a big deal, probably most of the workspaces are empty most of the time)

3. You have to describe use-cases for each menubar style (like for example, it's unclear to me why the indication of the currently active workspace is not enough)
The current approach might not be enough given that the text is quite thin and might take longer to identify in a quick glance, also the * on the currently active workspace is quite hard to catch too.

Providing an image type, and setting different visual states like: 
- bordered square (currently displayed space in a monitor but not active)
- filled square (currently active workspace + monitor) 
- bordered circle (for the active mode) 
This provides a more defined state of what’s going on with the state of the workspaces in a simple way

4. You have to take into consideration the fact that MacBook notch doesn't allow the indicator being too wide
This should not be a problem, regarding the image space vs the text space the menu will have almost the same size in the image. Also if the user chose the version that shows extra spaces, they accept the trade of in size, as a notch MacBook user, you can remove unnecessary items into the control center menu (again this is just UI customization of the OS, user should prioritize what they want there what’s useful to have there present all the time) also might be some users that will prefer not even displaying anything, like no menu at all, and the app is not accommodating for that neither, right?

5. You have to optimize number of possible menubar styles (make it as small as possible)
The number of styles are pretty small:
- text: it’s the default current one in the app
- image: instead of text use a representation of the character using an image.
- image-with-background-workspaces: same as image + text (to display the other “background” workspaces)

6. For every knob and configuration option you provide, you have to explain the importance in terms of workflows and use-cases
The use case for this config is:
- text: it’s the default current one in the app
- image: as a user I will prefer a more visual and identifiable clue of the active workspace, and the workspace in the other monitors, so it’s easier to quickly know at first glance my current state.
- image-with-background-workspaces: as a user with pretty bad memory I like to know which apps are open already or other spaces that have some window in it. Also another use case is when I close the lid of my MacBook and I open it the next day at the office, I like at a glance see my other workspaces to see which app I left open already and jump directly to that.

7. Regarding this: “Another observation: macOS native Spaces don't provide any indication of the currently active Space and people even fine with that…” they do display after you do 4 fingers up gesture and you move the mouse up to the top bar, they show the content of each desktop. Or another example is the “cmd + TAB” that will display all open apps. Or the small dot under each app icon on the dock. There are plenty of visual clues on macOS. Also newer versions of macOS provides for example Stage Manager functionality, that shows a bar in the side with a preview of all the open apps in each “desktop”.

**Reply by mobile-ar** [reply]
I updated this proposal to reflect more on the functionality I feel it was missing from the app than the UI change. But I still feel that the UI is more consistent with this images that share size and will make the menu size more consistent.

**Comment by mobile-ar** [comment, marked as answer]
This feature was added on `release in 0.19.0` as part of the Experimental UI settings menu option.
Also got improved with extra option lately.
More info in closed issue #711
